panic: runtime error: coffer partition disappears

**Describe the bug**

[This example](https://github.com/awnumar/memguard/tree/master/examples/deadlock) crashes sporadically on my system.

**To Reproduce**

`go test -v -count=10 ./examples/deadlock`

**Traces**

```
=== RUN   TestPanicsPoC
[] [14 227 172 214 47 31 152 84 236 122 1 107 223 116 37 0 158 52 188 103 100 78 201 60 202 48 122 169 17 135 212 236]
panic: open verify fail

goroutine 26 [running]:
github.com/awnumar/memguard/examples/deadlock.immediateOpen(0x59ed20, 0xc000054420, 0xc00000e060, 0xc000014440, 0x20, 0x20)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:80 +0x26e
github.com/awnumar/memguard/examples/deadlock.OpenEnclave.func1(0xc000010230, 0xa, 0xa, 0xc00001c260, 0xc000136000, 0xa, 0xa, 0xa, 0x59ed20, 0xc000054420)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:40 +0xe8
created by github.com/awnumar/memguard/examples/deadlock.OpenEnclave
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:32 +0x2bf
FAIL	github.com/awnumar/memguard/examples/deadlock	0.010s
FAIL
```
```
=== RUN   TestPanicsPoC
panic: runtime error: index out of range [0] with length 0

goroutine 266 [running]:
github.com/awnumar/memguard/core.(*Coffer).View(0xc0001082a0, 0x0, 0x0, 0x0)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/core/coffer.go:123 +0x207
github.com/awnumar/memguard/core.Open(0xc0000a80c0, 0x0, 0x0, 0x0)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/core/enclave.go:101 +0x60
github.com/awnumar/memguard.(*Enclave).Open(0xc0000a2038, 0x0, 0x0, 0x0)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/enclave.go:43 +0x32
github.com/awnumar/memguard/examples/deadlock.openVerify(0xc0000a2038, 0xc0000a6240, 0x20, 0x20, 0x0, 0x0)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:55 +0x58
github.com/awnumar/memguard/examples/deadlock.immediateOpen.func1(0xc0000a2038, 0xc0000a6240, 0x20, 0x20, 0xc000180960)
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:71 +0x49
created by github.com/awnumar/memguard/examples/deadlock.immediateOpen
	/home/awn/projects/src/go/src/github.com/awnumar/memguard/examples/deadlock/poc.go:70 +0xc8
FAIL	github.com/awnumar/memguard/examples/deadlock	0.068s
```

**System (please complete the following information):**
 - OS and Kernel Versions: `Linux 5.6.15-300.fc32.x86_64`
 - Memguard Version: `master`
 - Go Version: `go version go1.14.3 linux/amd64`

--- a/core/buffer.go
+++ b/core/buffer.go
@@ -23,7 +23,7 @@ Buffer is a structure that holds raw sensitive data.
 The number of Buffers that can exist at one time is limited by how much memory your system's kernel allows each process to mlock/VirtualLock. Therefore you should call DestroyBuffer on Buffers that you no longer need, ideally defering a Destroy call after creating a new one.
 */
 type Buffer struct {
-	sync.RWMutex // Local mutex lock
+	sync.RWMutex // Local mutex lock // TODO: this does not protect 'data' field
 
 	alive   bool // Signals that destruction has not come
 	mutable bool // Mutability state of underlying memory
@@ -44,12 +44,10 @@ NewBuffer is a raw constructor for the Buffer object.
 func NewBuffer(size int) (*Buffer, error) {
 	var err error
 
-	// Return an error if length < 1.
 	if size < 1 {
 		return nil, ErrNullBuffer
 	}
 
-	// Declare and allocate
 	b := new(Buffer)
 
 	// Allocate the total needed memory
@@ -119,18 +117,14 @@ func (b *Buffer) Freeze() {
 }
 
 func (b *Buffer) freeze() error {
-	// Attain lock.
 	b.Lock()
 	defer b.Unlock()
 
-	// Check if destroyed.
 	if !b.alive {
 		return nil
 	}
 
-	// Only do anything if currently mutable.
 	if b.mutable {
-		// Make the memory immutable.
 		if err := memcall.Protect(b.inner, memcall.ReadOnly()); err != nil {
 			return err
 		}
@@ -148,18 +142,14 @@ func (b *Buffer) Melt() {
 }
 
 func (b *Buffer) melt() error {
-	// Attain lock.
 	b.Lock()
 	defer b.Unlock()
 
-	// Check if destroyed.
 	if !b.alive {
 		return nil
 	}
 
-	// Only do anything if currently immutable.
 	if !b.mutable {
-		// Make the memory mutable.
 		if err := memcall.Protect(b.inner, memcall.ReadWrite()); err != nil {
 			return err
 		}
@@ -176,7 +166,6 @@ func (b *Buffer) Scramble() {
 }
 
 func (b *Buffer) scramble() error {
-	// Attain lock.
 	b.Lock()
 	defer b.Unlock()
 	return Scramble(b.Data())
@@ -269,8 +258,6 @@ func (l *bufferList) add(b ...*Buffer) {
 	l.Lock()
 	defer l.Unlock()
 
-	// fmt.Printf("\n\nThere are %d buffers\n\n\n", len(buffers.list))
-
 	l.list = append(l.list, b...)
 }
 
@@ -289,8 +276,6 @@ func (l *bufferList) copy() []*Buffer {
 func (l *bufferList) remove(b *Buffer) {
 	l.Lock()
 	defer l.Unlock()
-
-	// fmt.Printf("\n\nThere are %d buffers\n\n\n", len(buffers.list))
 
 	for i, v := range l.list {
 		if v == b {

--- a/core/coffer.go
+++ b/core/coffer.go
@@ -16,33 +16,27 @@ var ErrCofferExpired = errors.New("<memguard::core::ErrCofferExpired> attempted 
 Coffer is a specialized container for securing highly-sensitive, 32 byte values.
 */
 type Coffer struct {
-	sync.RWMutex
+	sync.Mutex
 
-	left  *Buffer // Left partition.
-	right *Buffer // Right partition.
+	left  *Buffer
+	right *Buffer
 
-	rand *Buffer // Static allocation for fast random bytes reading.
+	rand *Buffer
 }
 
 // NewCoffer is a raw constructor for the *Coffer object.
 func NewCoffer() *Coffer {
-	// Create a new Coffer object.
 	s := new(Coffer)
-
-	// Allocate the partitions.
 	s.left, _ = NewBuffer(32)
 	s.right, _ = NewBuffer(32)
 	s.rand, _ = NewBuffer(32)
 
-	// Initialise with a random 32 byte value.
-	s.Initialise()
+	s.Init()
 
 	go func(s *Coffer) {
-		for {
-			// Sleep for the specified interval.
-			time.Sleep(interval)
+		ticker := time.NewTicker(interval)
 
-			// Re-key the contents, exiting the routine if object destroyed.
+		for range ticker.C {
 			if err := s.Rekey(); err != nil {
 				break
 			}
@@ -52,27 +46,15 @@ func NewCoffer() *Coffer {
 	return s
 }
 
-/*
-Initialise is used to reset the value stored inside a Coffer to a new random 32 byte value, overwriting the old.
-*/
-func (s *Coffer) Initialise() error {
-	// Check if it has been destroyed.
+// Init is used to reset the value stored inside a Coffer to a new random 32 byte value, overwriting the old.
+func (s *Coffer) Init() error {
+	s.Lock()
+	defer s.Unlock()
+
 	if s.Destroyed() {
 		return ErrCofferExpired
 	}
 
-	if err := s.initialise(); err != nil {
-		Panic(err)
-	}
-	return nil
-}
-
-func (s *Coffer) initialise() error {
-	// Attain the mutex.
-	s.Lock()
-	defer s.Unlock()
-
-	// Overwrite the old value with fresh random bytes.
 	if err := Scramble(s.left.Data()); err != nil {
 		return err
 	}
@@ -94,40 +76,23 @@ func (s *Coffer) initialise() error {
 View returns a snapshot of the contents of a Coffer inside a Buffer. As usual the Buffer should be destroyed as soon as possible after use by calling the Destroy method.
 */
 func (s *Coffer) View() (*Buffer, error) {
-	//fmt.Printf("\n\nThere are %d buffers\n\n\n", len(buffers.list))
+	s.Lock()
+	defer s.Unlock()
 
-	// Check if it's destroyed.
 	if s.Destroyed() {
 		return nil, ErrCofferExpired
 	}
-	//fmt.Printf("\n\n%s\n\n\n", "s is not destroyed")
 
-	// Create a new Buffer for the data.
 	b, _ := NewBuffer(32)
-
-	//fmt.Printf("\n\n%s\n\n\n", "made buffer")
-
-	// Attain a read-only lock.
-	s.RLock()
-	defer s.RUnlock()
-
-	//fmt.Printf("\n\n%s\n\n\n", "got lock")
-
-	//fmt.Println(s.right.Data(), s.left.Data())
 
 	// data = hash(right) XOR left
 	h := Hash(s.right.Data())
 
-	//fmt.Printf("\n\n%s\n\n\n", "got hash")
 	for i := range b.Data() {
 		b.Data()[i] = h[i] ^ s.left.Data()[i]
 	}
-	//fmt.Printf("\n\n%s\n\n\n", "computed plaintext")
 	Wipe(h)
 
-	// fmt.Printf("\n\n%s\n\n\n", "got data")
-
-	// Return the view.
 	return b, nil
 }
 
@@ -135,23 +100,13 @@ func (s *Coffer) View() (*Buffer, error) {
 Rekey is used to re-key a Coffer. Ideally this should be done at short, regular intervals.
 */
 func (s *Coffer) Rekey() error {
-	// Check if it has been destroyed.
+	s.Lock()
+	defer s.Unlock()
+
 	if s.Destroyed() {
 		return ErrCofferExpired
 	}
 
-	if err := s.rekey(); err != nil {
-		Panic(err)
-	}
-	return nil
-}
-
-func (s *Coffer) rekey() error {
-	// Attain the mutex.
-	s.Lock()
-	defer s.Unlock()
-
-	// Attain 32 bytes of fresh cryptographic buf32.
 	if err := Scramble(s.rand.Data()); err != nil {
 		return err
 	}
@@ -177,28 +132,18 @@ func (s *Coffer) rekey() error {
 /*
 Destroy wipes and cleans up all memory related to a Coffer object. Once this method has been called, the Coffer can no longer be used and a new one should be created instead.
 */
-func (s *Coffer) Destroy() {
-	if err := s.destroy(); err != nil {
-		Panic(err)
-	}
-}
-
-func (s *Coffer) destroy() error {
-	// Attain the mutex.
+func (s *Coffer) Destroy() error {
 	s.Lock()
 	defer s.Unlock()
 
-	// Destroy the partitions.
 	err1 := s.left.destroy()
 	if err1 == nil {
 		buffers.remove(s.left)
 	}
-
 	err2 := s.right.destroy()
 	if err2 == nil {
 		buffers.remove(s.right)
 	}
-
 	err3 := s.rand.destroy()
 	if err3 == nil {
 		buffers.remove(s.rand)
@@ -222,8 +167,5 @@ func (s *Coffer) destroy() error {
 
 // Destroyed returns a boolean value indicating if a Coffer has been destroyed.
 func (s *Coffer) Destroyed() bool {
-	s.RLock()
-	defer s.RUnlock()
-
 	return s.left.data == nil || s.right.data == nil
 }

--- a/core/coffer_test.go
+++ b/core/coffer_test.go
@@ -113,10 +113,7 @@ func TestCofferView(t *testing.T) {
 func TestCofferRekey(t *testing.T) {
 	s := NewCoffer()
 
-	// Halt the rekey cycle.
-	s.Lock()
-
-	// Remember the value stored inside.
+	// remember the value stored inside
 	view, err := s.View()
 	if err != nil {
 		t.Error("unexpected error;", err)
@@ -125,18 +122,16 @@ func TestCofferRekey(t *testing.T) {
 	copy(orgValue, view.Data())
 	view.Destroy()
 
-	// Remember the value of the partitions.
+	// remember the value of the partitions
 	left := make([]byte, 32)
 	right := make([]byte, 32)
+	s.Lock() // halt re-key cycle
 	copy(left, s.left.Data())
 	copy(right, s.right.Data())
+	s.Unlock() // un-halt re-key cycle
 
-	// Manually re-key before we continue.
-	s.Unlock()
-	s.Rekey()
-	s.Lock()
+	s.Rekey() // force a re-key
 
-	// Get another view of the contents.
 	view, err = s.View()
 	if err != nil {
 		t.Error("unexpected error;", err)
@@ -145,17 +140,14 @@ func TestCofferRekey(t *testing.T) {
 	copy(newValue, view.Data())
 	view.Destroy()
 
-	// Compare the values.
 	if !bytes.Equal(orgValue, newValue) {
 		t.Error("value inside coffer changed!!")
 	}
 
-	// Compare the partition values.
 	if bytes.Equal(left, s.left.Data()) || bytes.Equal(right, s.right.Data()) {
 		t.Error("partition values did not change")
 	}
 
-	s.Unlock() // Release lock to allow destruction.
 	s.Destroy()
 
 	if err := s.Rekey(); err != ErrCofferExpired {

--- a/core/coffer_test.go
+++ b/core/coffer_test.go
@@ -9,7 +9,7 @@ func TestNewCoffer(t *testing.T) {
 	s := NewCoffer()
 
 	// Attain a lock to halt the verify & rekey cycle.
-	s.RLock()
+	s.Lock()
 
 	// Verify that fields are not nil.
 	if s.left == nil || s.right == nil {
@@ -32,11 +32,11 @@ func TestNewCoffer(t *testing.T) {
 		t.Error("right side is zeroed")
 	}
 
-	s.RUnlock() // Release mutex to allow destruction.
+	s.Unlock() // Release mutex to allow destruction.
 	s.Destroy()
 }
 
-func TestCofferInitialise(t *testing.T) {
+func TestCofferInit(t *testing.T) {
 	s := NewCoffer()
 
 	// Get the value stored inside.
@@ -48,8 +48,8 @@ func TestCofferInitialise(t *testing.T) {
 	copy(value, view.Data())
 	view.Destroy()
 
-	// Re-initialise the buffer with a new value.
-	if err := s.Initialise(); err != nil {
+	// Re-init the buffer with a new value.
+	if err := s.Init(); err != nil {
 		t.Error("unexpected error;", err)
 	}
 
@@ -70,7 +70,7 @@ func TestCofferInitialise(t *testing.T) {
 	s.Destroy()
 
 	// Check error condition.
-	if err := s.Initialise(); err != ErrCofferExpired {
+	if err := s.Init(); err != ErrCofferExpired {
 		t.Error("expected ErrCofferExpired; got", err)
 	}
 }
@@ -114,7 +114,7 @@ func TestCofferRekey(t *testing.T) {
 	s := NewCoffer()
 
 	// Halt the rekey cycle.
-	s.RLock()
+	s.Lock()
 
 	// Remember the value stored inside.
 	view, err := s.View()
@@ -132,9 +132,9 @@ func TestCofferRekey(t *testing.T) {
 	copy(right, s.right.Data())
 
 	// Manually re-key before we continue.
-	s.RUnlock()
+	s.Unlock()
 	s.Rekey()
-	s.RLock()
+	s.Lock()
 
 	// Get another view of the contents.
 	view, err = s.View()
@@ -155,7 +155,7 @@ func TestCofferRekey(t *testing.T) {
 		t.Error("partition values did not change")
 	}
 
-	s.RUnlock() // Release lock to allow destruction.
+	s.Unlock() // Release lock to allow destruction.
 	s.Destroy()
 
 	if err := s.Rekey(); err != ErrCofferExpired {

--- a/core/enclave.go
+++ b/core/enclave.go
@@ -1,17 +1,27 @@
 package core
 
-import "errors"
+import (
+	"errors"
+	"sync/atomic"
+	"unsafe"
+)
 
 var (
-	// Declare a key for use in encrypting data this session.
-	key *Coffer
+	key unsafe.Pointer // *Coffer
 )
 
 func init() {
-	// Initialize the key declared above with a random value.
-	if key == nil {
-		key = NewCoffer()
+	if swapped := atomic.CompareAndSwapPointer(&key, nil, nil); swapped {
+		setKey(NewCoffer())
 	}
+}
+
+func getKey() *Coffer {
+	return (*Coffer)(atomic.LoadPointer(&key))
+}
+
+func setKey(k *Coffer) {
+	atomic.StorePointer(&key, unsafe.Pointer(k))
 }
 
 // ErrNullEnclave is returned when attempting to construct an enclave of size less than one.
@@ -37,7 +47,7 @@ func NewEnclave(buf []byte) (*Enclave, error) {
 	e := new(Enclave)
 
 	// Get a view of the key.
-	k, err := key.View()
+	k, err := getKey().View()
 	if err != nil {
 		return nil, err
 	}
@@ -98,7 +108,7 @@ func Open(e *Enclave) (*Buffer, error) {
 	}
 
 	// Grab a view of the key.
-	k, err := key.View()
+	k, err := getKey().View()
 	if err != nil {
 		return nil, err
 	}

--- a/core/enclave_test.go
+++ b/core/enclave_test.go
@@ -10,7 +10,7 @@ func TestEnclaveInit(t *testing.T) {
 		t.Error("key is nil")
 	}
 
-	view, err := key.View()
+	view, err := getKey().View()
 	if err != nil {
 		t.Error("unexpected error;", err)
 	}

--- a/core/exit.go
+++ b/core/exit.go
@@ -19,8 +19,9 @@ func Purge() {
 
 	func() {
 		// Halt the re-key cycle and prevent new enclaves.
-		key.Lock()
-		defer key.Unlock()
+		k := getKey()
+		k.Lock()
+		defer k.Unlock()
 
 		// Get a snapshot of existing Buffers.
 		snapshot := buffers.flush()
@@ -49,9 +50,8 @@ func Purge() {
 		}
 	}()
 
-	// Destroy and recreate the key.
-	key.Destroy() // should be a no-op
-	key = NewCoffer()
+	// Create a new key.
+	setKey(NewCoffer())
 
 	// If we encountered an error, panic.
 	if opErr != nil {
@@ -64,7 +64,7 @@ Exit terminates the process with a specified exit code but securely wipes and cl
 */
 func Exit(c int) {
 	// Wipe the encryption key used to encrypt data inside Enclaves.
-	key.Destroy()
+	getKey().Destroy()
 
 	// Get a snapshot of existing Buffers.
 	snapshot := buffers.copy() // copy ensures the buffers stay in the list until they are destroyed.

--- a/core/exit_test.go
+++ b/core/exit_test.go
@@ -16,11 +16,9 @@ func TestPurge(t *testing.T) {
 		t.Error(err)
 	}
 
-	// Keep a reference to the old key.
-	oldKey := key
-
-	// Purge the session.
+	oldKey := getKey()
 	Purge()
+	key := getKey()
 
 	// Verify that the buffers list contains only the important buffers.
 	buffers.RLock()

--- a/examples/deadlock/poc.go
+++ b/examples/deadlock/poc.go
@@ -11,9 +11,9 @@ import (
 	"time"
 
 	"github.com/awnumar/memguard"
+	"lukechampine.com/frand"
 )
 
-// OpenEnclave ...
 func OpenEnclave(ctx context.Context) {
 	n := 10
 	data := make([][]byte, n)
@@ -21,25 +21,23 @@ func OpenEnclave(ctx context.Context) {
 	for i := range data {
 		data[i] = make([]byte, 32)
 		buf := make([]byte, 32)
-		io.ReadFull(rand.Reader, buf)
+		if _, err := io.ReadFull(rand.Reader, buf); err != nil {
+			panic("failed to read random data")
+		}
 		copy(data[i], buf)
 		enclaves[i] = memguard.NewEnclave(buf)
 	}
 
 	threads := 20
 	for i := 0; i < threads; i++ {
-		j := 0
 		go func(ctx context.Context) {
 			for {
 				select {
 				case <-ctx.Done():
 					return
 				default:
-					{
-						//fmt.Printf("open enclave %d \n", j)
-						immediateOpen(ctx, enclaves[j], data[j])
-						j = (j + 1) % n
-					}
+					j := frand.Intn(n)
+					immediateOpen(ctx, enclaves[j], data[j])
 				}
 			}
 		}(ctx)
@@ -65,27 +63,17 @@ func openVerify(lock *memguard.Enclave, exp []byte) error {
 }
 
 func immediateOpen(ctx context.Context, lock *memguard.Enclave, exp []byte) {
-	start := time.Now()
 	c1 := make(chan error, 1)
 	go func() {
 		err := openVerify(lock, exp)
 		c1 <- err
 	}()
-	var dur time.Duration
+
 	select {
 	case err := <-c1:
-		{
-			dur = time.Since(start)
-			if err != nil {
-				panic(err)
-			}
+		if err != nil {
+			panic(err)
 		}
 	case <-ctx.Done():
-		{
-			dur = time.Since(start)
-			fmt.Printf("### timeout \n")
-		}
 	}
-	//fmt.Printf("%d, %d \n", start.UnixNano(), dur.Nanoseconds())
-	_ = dur
 }

--- a/examples/deadlock/poc_test.go
+++ b/examples/deadlock/poc_test.go
@@ -9,7 +9,7 @@ import (
 	"time"
 )
 
-const duration = 2 * time.Minute
+const duration = 1 * time.Minute
 
 func TestPanicsPoC(t *testing.T) {
 	sigs := make(chan os.Signal, 1)

--- a/go.mod
+++ b/go.mod
@@ -5,6 +5,10 @@ go 1.18
 require (
 	github.com/awnumar/memcall v0.1.2
 	golang.org/x/crypto v0.0.0-20220525230936-793ad666bf5e
+	lukechampine.com/frand v1.4.2
 )
 
-require golang.org/x/sys v0.0.0-20210615035016-665e8c7367d1 // indirect
+require (
+	github.com/aead/chacha20 v0.0.0-20180709150244-8b13a72661da // indirect
+	golang.org/x/sys v0.0.0-20210615035016-665e8c7367d1 // indirect
+)

